### PR TITLE
fix(facade): MapWrapper.createFromPairs

### LINKED ISSUE #1640: Safari's ES6 Map does not support constructor params, we need a better facade for it.
Open hello world in Safari (tested 7 and 8), the following error appears `TypeError: Map constructor does not accept arguments`.

### PR DESCRIPTION
**Pull request by chirayuk**
Closes #1640

### CODE REVIEW

File: modules/angular2/src/facade/collection.ts [hunk 1]
```diff
@@ -10,6 +10,24 @@ export var Map = global.Map;
...
+    if (new Map([1, 2]).size === 2) {
+      return function createMapFromPairs(pairs: List<any>): Map<any, any> {
+        return new Map(pairs);
+      };
+    }
+  } catch (e) {
+  }
+  return function createMapAndPopulateFromPairs(pairs: List<any>): Map<any, any> {
+    var map = new Map();
+    pairs.forEach(function(value, key) { map.set(key, value); });
```
Comment (mhevery) [inline modules/angular2/src/facade/collection.ts]: Please don't use forEach, see: https://github.com/angular/angular/wiki/performance#foreach

Comment (chirayuk) [inline modules/angular2/src/facade/collection.ts]: Changed to use a for-loop instead of forEach.  Thanks!

**Comment by chirayuk** [comment]
Travis is now passing.  (I have trouble running tests locally.)  Good to merge?

File: modules/angular2/src/facade/collection.ts [hunk 2]
```diff
@@ -10,6 +10,27 @@ export var Map = global.Map;
...
+var createMapFromPairs: {(pairs: List<any>): Map<any, any>} = (function() {
+  try {
+    if (new Map([1, 2]).size === 2) {
+      return function createMapFromPairs(pairs: List<any>): Map<any, any> {
+        return new Map(pairs);
+      };
+    }
+  } catch (e) {
+  }
+  return function createMapAndPopulateFromPairs(pairs: List<any>): Map<any, any> {
```
Comment (vicb) [inline modules/angular2/src/facade/collection.ts]: should this be moved inside the catch ?

**Comment by vicb** [comment]
`createMapFromPairs` seems to be called when the module is imported. 
An alternative would be to call on on first `MapWrapper.createMapFromPairs()`.

Otherwise LGTM.

File: modules/angular2/src/facade/collection.ts (hunk 2 above)
Comment (chirayuk) [inline modules/angular2/src/facade/collection.ts]: I like it outside the catch.  For style check reasons, the empty catch is required to occupy two lines :)

While the browser I tested throw an exception when passed the parameter, it's possible that some polyfill or a different browser just ignores the parameter.  That's why I have the size check.  Failing the size check also requires using the slower approach just like for the catch clause.  In the interest of DRY, I have a single return as a catch all right at the end to accomplish this.

**Comment by chirayuk** [comment]
@vicb I get what you're saying.  However, the call is exactly once on first import to determine which which function should be used as opposed to being done lazily.  If the application uses that constructor even once (which is true for us), then the eager approach is totally fine and has the advantage of simplicity and avoids checks at actual usage times.

**Comment by vicb** [comment]
Thanks for the explanation. Makes sense.

**Comment by chirayuk** [comment]
Landed as af9dcad8e347a4c82874feab6f8f9c44df7da499

**Pull request closed without merging**